### PR TITLE
Add eval script

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
### Description of the change

- Adds a script for model evaluation for generation based classification.

Sample usage is as follows (details can be seen with `--help` - this is is just for the sake of example and illustrates how to run something). Note that `train` is the default split for loading a local file, and the usage of `--split` matches that of a normal HF dataset. A `--delimiter` may also be passed for multi output classification.
```
python3 run_evaluation.py --model TinyLlama/TinyLlama-1.1B-step-50K-105b  --data_path stanford_alpaca/alpaca_data.json  --max_new_tokens 10  --purge_results --split train[:50] --delimiter , 
```

Running an evaluation produces a directory containing 4 files, each of which is shown with an example snippet below.

1. `eval_metrics.json` - contains the metrics themselves.
```json
{
    "accuracy": 0.02,
    "f1": {
        "macro": 0.0196078431372549,
        "micro": 0.02
    },
    "precision": {
        "macro": 0.5,
        "micro": 0.02
    },
    "recall": {
        "macro": 0.02,
        "micro": 0.02
    }
}
```

2. `experiment_metadata.json` - contains misc info about the experiment that might be useful later.
```json
{
    "data_path": "stanford_alpaca/alpaca_data.json",
    "max_new_tokens": 10,
    "model": "TinyLlama/TinyLlama-1.1B-step-50K-105b"
}
```

3. `label_map.json` - contains the mapping from integer IDs to `str` classes based on the ground truth target sequences.  This is mostly useful for understanding the metrics if they look incorrect, because it's what is actually being used in the f1/accuracy calculations. If something is generated that is not in the ground truth sequences, it's put into a single `<UNKNOWN LABEL>` garbage class.

Note that the labels of this file do not matter - this is essentially garbage from an untuned model and just for the sake of example.
```json
{
    "0": "",
    "1": "![](https://github.com/",
    "2": "![logo](https://user-images",
    "3": "### Instruction:\nGenerate",
    ...
    "44": "<UNKNOWN LABEL>"
}
```

4. `preds_and_references.json` - contains the formatted inputs provided to the model, the predicted target texts, and the ground truth target texts. This is also primarily for debugging purposes, in case things look a bit strange in the metrics.

```json
[
    {
        "formatted input": "Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\nGive three tips for staying healthy.\n\n### Response:",
        "predicted target": "### Instruction:\nGive",
        "ref target": "1.Eat a balanced diet and make sure to include plenty of fruits and vegetables. \n2. Exercise regularly to keep your body active and strong. \n3. Get enough sleep and maintain a consistent sleep schedule."
    },
    {
        "formatted input": "Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\nWhat are the three primary colors?\n\n### Response:",
        "predicted target": "### Instruction:\nWhat is",
        "ref target": "The three primary colors are red, blue, and yellow."
    },
    ...
]
```

Note that currently this adds sklearn as a dependency to the project - we probably don't want to do this since scripts are not distributed, but we should figure out the right way to structure scripts dependencies. Or maybe we should just add it to dev dependencies in `pyproject.toml`?

### How to verify the PR
See script in comments.

### Was the PR tested

- [ ] I have added >=1 unit test(s) for every new method I have added.
- [ ] I have ensured all unit tests pass

### CODE REVIEW

**Comment by alex-jw-brooks** [comment]
# How to Validate

There are a couple of cases to validate for metric correctness - the easiest way to do this is to directly test the already delimited strings. To make it easier, the example is taken from this [blog post](https://medium.com/synthesio-engineering/precision-accuracy-and-f1-score-for-multi-label-classification-34ac6bdfb404).

### No Unknown Class
This is exactly the same as described in the blog post.
```python3
import json
from run_evaluation import map_predictions_and_references_to_encoded_vectors, compute_metrics_dict_multi

references = [
    ["cat", "dog"],
    ["cat", "bird"],
    ["bird"],
    [],
]
predictions = [
    ["dog", "bird"],
    ["cat"],
    ["bird"],
    ["cat", "bird"],
]

(
    pred_vecs,
    ref_vecs,
    eval_label_map,
) = map_predictions_and_references_to_encoded_vectors(predictions, references)
metrics_dict = compute_metrics_dict_multi(pred_vecs, ref_vecs)
print(json.dumps(metrics_dict, indent=4, sort_keys=True))
```
Produces:
```json
{
    "accuracy": 0.25,
    "f1": {
        "macro": 0.6333333333333333,
        "micro": 0.5454545454545454
    },
    "precision": {
        "macro": 0.611111111111111,
        "micro": 0.5
    },
    "recall": {
        "macro": 0.6666666666666666,
        "micro": 0.6
    }
}
```
Note that .63 / .55 match for the micro and macro f1 - since accuracy in this case is checking complete match, .25 is correct.


### Unknown Class
Any garbage values should be collapsed into a *single* unknown value for that instance. Change the predictions variable in the script above from:
```python3
predictions = [
    ["dog", "bird"],
    ["cat"],
    ["bird"],
    ["cat", "bird"],
]
```
to:
```python3
predictions = [
    ["dog", "bird"],
    ["cat"],
    ["bird"],
    ["cat", "bird", "garbage1", "garbage2", "garbage3"],
]
```
(the last of which effectively resolves to `["cat", "bird", "<UNKNOWN>"]` during encoding.

**Review by Ssukriti — APPROVED** [review]

**Review by Ssukriti — APPROVED** [review]
will test it out with actual examples today, merging for now

**Comment by Ssukriti** [comment]
thanks for the detail description on how to validate :)
